[New Tutorial] Create a simple Voting dApp on ShimmerEVM (#909)
iota-wiki/iota-wiki: tutorials/pages/shimmerevm-dapp-voting.md

The first thing you need to do is check that Metamask is accessible from your dApp:
```javascript!
const provider = new ethers.providers.Web3Provider(window.ethereum);
```
If you're able to access `provider`, it's all good. Let's proceed.

:::info
Learn more about what else you can do with ethers from the [ethers docs](https://docs.ethers.org/v5/getting-started/).
:::

You need to create a `signer` which will be used to sign transactions from your dApp:
```javascript!
const signer = provider.getSigner();
```

Create a contract client which will represent the contract and help with calling functions of it using `CONTRACT_ADDRESS` and `CONTRACT_ABI`, which you should have from the above steps:

```javascript!
const contractClient = new ethers.Contract(
    CONTRACT_ADDRESS,
    CONTRACT_ABI,
    signer
);
```
Use the `contractClient` for all fetch methods, such as fetching proposals:
`await contractClient.getProposals(i)`.
However, to use state-changing methods, you need a `contractSigner` instance, like:
```javascript!
const contractSigner = contractClient.connect(signer);
```

> Now, for instance, to vote, you can assign the action to a button and simply call:

Reviewer: ```suggestion
Now you can assign the `contractSigner.vote` action to a button and call:
```

```javascript!
let tx = contractSigner.vote(i);
```

> where `i` is the index of the proposal you want to vote for.

Reviewer: ```suggestion

* Where `i` is the index of the proposal you want to vote for.
```

> ## Result

Reviewer: ```suggestion
## Result

```

If you followed everything above, it should look something like this:
![Voting Demo](../static/voting-dapp.jpg "Voting dApp UI")

> Feel free to make changes to the design to suit your needs.

Reviewer: ```suggestion
```

## Requirements

You need the following before you get started:

- **Network**: A network where you will deploy and interact with the smart contract. In this case, the ShimmerEVM.
- **Wallet**: You will need an interface to connect to the network, verify your identity, and pay the transaction fees. You can use any wallet you choose; this tutorial will use [Metamask](https://metamask.io/), but you can use any wallet you are comfortable with.
- **Smart Contract**: You will need a smart contract to deploy on-chain to handle all state records on-chain and any changes made to them.
- **Code Editor or IDE**: Have an IDE or any code editor ready to start working.
- **Client Library**: This tutorial will use [Ethers](https://docs.ethers.org/v5/). But you can use any other library that you feel more comfortable in.

## Project Setup

This tutorial will use a vanilla js project. However, most of it should still work on any js framework like react, vue, sveltekit, etc.

## Create a Smart Contracts in Solidity

This tutorial will use a simple Voting Smart Contract, that will be called `Voting.sol`.

#### Create and map the structs

The first thing you need is a `struct` called `Voter`:

> ```solidity!

Author: It adds line number. Looks better and also useful for reference.

> ```solidity!

Reviewer 2: Ohh. Didn't know that 🤔 Got it

> ```solidity!

Author: [Hackmd](https://hackmd.io/@anistark/shimmerevm-dapp) for reference.
Now that you mention, not sure if it'll work in docusaurus. 🥲

    address delegate; // person delegated to
    uint vote;   // index of the voted proposal
}
```

Once you have created the `Voter`, you will need `Proposal` `struct` to nominate who to vote on:

```solidity!
mapping(address => Voter) public voters;
```

Once you have created the `Voter`, you will need `Proposal` `struct` to nominate who to vote on:

```solidity!
struct Proposal {
    bytes32 name;   // short name (up to 32 bytes)
    uint voteCount; // number of accumulated votes
}
```

:::tip Gas Fees
If you can, limit the length to a certain number of bytes. You must cover gas fees for any data you store on the chain if you use one of bytes1 to bytes32 as they are much cheaper. In other words, if you use short names, your gas fees will be lower.

:::

Next, you need to add a vote function, which will register the vote of every user toward a proposal. Let's start with the following:

```solidity!
function vote(uint proposal) public {
    Voter storage sender = voters[msg.sender];
}
```

As you need to check if the voter is allowed to vote, you should add the following:

```solidity!
require(sender.weight != 0, "Has no right to vote");
```

> You also need to check if the voter has already voted, so you should also add the following conditional :

Reviewer 2: ```suggestion
You also need to check if the voter has already voted, so you should also add the following conditional:
```

:::info Ethers

Learn more about what else you can do with ethers from the [ethers docs](https://docs.ethers.org/v5/getting-started/).

:::

You need to create a `signer` which will be used to sign transactions from your dApp:

```javascript!
const signer = provider.getSigner();
```

Create a contract client which will represent the contract and help with calling the contract's functions using `CONTRACT_ADDRESS` and `CONTRACT_ABI`, which you should have from the above steps:

```javascript!
const contractClient = new ethers.Contract(
    CONTRACT_ADDRESS,
    CONTRACT_ABI,
    signer
);
```

Use the `contractClient` for all fetch methods, such as fetching proposals:

```js
await contractClient.getProposals(i);
```

If you want to use state-changing methods, you need to create a `contractSigner` instance:

```javascript!
const contractSigner = contractClient.connect(signer);
```

Now you can assign the `contractSigner.vote` action to a button and call:

```javascript!
let tx = contractSigner.vote(i);
```

- Where `i` is the index of the proposal you want to vote for.

## Result

If you followed everything above, it should look something like this:

> ![Voting Demo](../static/voting-dapp.jpg 'Voting dApp UI')

Reviewer 2: ```suggestion
![Voting Demo](/voting-dapp.jpg 'Voting dApp UI')
```

# Create a Simple Voting dApp on ShimmerEVM

## Goal

This tutorial will guide you as you write and deploy a voting dApp on ShimmerEVM. You can vote on a few options and view the results; all data will be stored on-chain. Please note **this tutorial is meant for learning purposes only **.

## Expected Outcome

By the end of this tutorial, you will be able to cast your vote from several available options and view the results. All the data is stored securely on the blockchain.

## Requirements

You need the following before you get started:

- **Network**: A network where you will deploy and interact with the smart contract. In this case, the ShimmerEVM.
- **Wallet**: You will need an interface to connect to the network, verify your identity, and pay the transaction fees. You can use any wallet you choose; this tutorial will use [Metamask](https://metamask.io/), but you can use any wallet you are comfortable with.
- **Smart Contract**: You will need a smart contract to deploy on-chain to handle all state records on-chain and any changes made to them.
- **Code Editor or IDE**: Have an IDE or any code editor ready to start working.

> - **Client Library**: This tutorial will use [Ethers](https://docs.ethers.org/v5/). But you can use any other library that you feel more comfortable in.

Author: I updated it. Removed code editor.

> - **Client Library**: This tutorial will use [Ethers](https://docs.ethers.org/v5/). But you can use any other library that you feel more comfortable in.

Reviewer 2: Ohh. Missed that somehow 🤔

## Project Setup

This tutorial will use a vanilla js project. However, most of it should still work on any js framework like react, vue, sveltekit, etc.

## Create a Smart Contracts in Solidity

This tutorial will use a simple Voting Smart Contract, that will be called `Voting.sol`.

#### Create and map the structs

The first thing you need is a `struct` called `Voter`:

```solidity!
struct Voter {
    uint weight; // weight is accumulated by delegation
    bool voted;  // if true, that person already voted
    address delegate; // person delegated to
    uint vote;   // index of the voted proposal
}
```

Once you have created the `Voter`, you will need `Proposal` `struct` to nominate who to vote on:

```solidity!
mapping(address => Voter) public voters;
```

> Once you have created the `Voter`, you will need `Proposal` `struct` to nominate who to vote on:

Author: Yes.

```solidity!
function winningProposal() public view returns (uint winningProposal_) {
    uint winningVoteCount = 0;
    for (uint p = 0; p < proposals.length; p++) {
        if (proposals[p].voteCount > winningVoteCount) {
            winningVoteCount = proposals[p].voteCount;
            winningProposal_ = p;
        }
    }
}
```

To make things a little more human readable, add a function `winnerName` that returns the name of the winner:

```solidity!
function winnerName() public view returns (bytes32 winnerName_) {
    winnerName_ = proposals[winningProposal()].name;
}
```

At this point, the basic functionalities of the smart contract are taken care of. However, you could also need a few more functionalities, such as delegating votes and giving someone else the right to vote. I have added those in the final code, so feel free to keep or remove them as you see fit.

Here's the complete smart contract code:

```solidity!=
// SPDX-License-Identifier: GPL-3.0
pragma solidity >=0.7.0 <0.9.0;

/**
 * @title Voting
 * @dev Implements voting process along with vote delegation
 */
contract Voting {

> 

Reviewer 2: Who do you want to target with this tutorial? 
If beginners then I guess you should also explain this part? If not ignore this comment :)
